### ISSUE TITLE
block transfer issues leshan <-> wakaama

### ISSUE THREAD

**Issue by qleisan** [eclipse/wakaama#510]
While testing #461 (block transfer)

Test with leshan server and wakaama client (leshan setting: multi-value: SENML_JSON, single-value: SENML-JSON)

"GUI write" to 31024/10/5 with ~200 bytes => "Request Entity Too Large" (leshan tries to use 512 byte block size?)

"GUI read" from 31024/10/5 with ~200 bytes => OK
128 byte block size used. (setup with "change /31024/10/5 <200bytes>")

"GUI read" from 31024/10/5 with 1024 bytes => "Internal server error"

**Comment by sbernard31** [comment]
Wireshark capture could help to see where is the issue.

> "GUI write" to 31024/10/5 with ~200 bytes => "Request Entity Too Large" (leshan tries to use 512 byte block size?)

We should investigate the reason of this (capture should help)
Californium should be able to handle this and adapt it's blocksize.

> "GUI read" from 31024/10/5 with ~200 bytes => OK

:+1:  (even for working one capture could be useful to see if standard is well respected)

> "GUI read" from 31024/10/5 with 1024 bytes => "Internal server error"

"Internal server error" as response code sent by Wakaama, right ?

**Comment by sbernard31** [comment]
(Note that "block1 negotiation" and "SenML-JSON" are fresh feature from Californium and Leshan so a bug could be present at that side too)

**Comment by qleisan** [comment]
Great input. Downloaded a newer version of leshan (github "Standalone Demos" "v2x"). There was no negotiation, but if I manually edited PREFERRED_BLOCK_SIZE= to "128" (prev "512") it worked fine.

**Comment by qleisan** [comment]
[201208_leshan_read_200.pcapng.zip](https://github.com/eclipse/wakaama/files/5661867/201208_leshan_read_200.pcapng.zip)

Uploading a successful "GUI read" from 31024/10/5 with ~200 bytes => OK"

NB. copy&paste "change /31024/10/5 <200bytes>" to wakaama cli seems to fail sometimes/often. When read back with dump /31024/10 there might be garbage characters at the end. If that is the case it seems that leshan will not succeed to do GET

**Comment by sbernard31** [comment]
> Uploading a successful "GUI read" from 31024/10/5 with ~200 bytes => OK"

I looked at the capture and it looks OK.

> NB. copy&paste "change /31024/10/5 <200bytes>" to wakaama cli seems to fail sometimes/often. When read back with dump /31024/10 there might be garbage characters at the end. If that is the case it seems that leshan will not succeed to do GET

I'm not sure to get that point. :thinking: 

> Downloaded a newer version of leshan (github "Standalone Demos" "v2x"). There was no negotiation,

I tried to understand the capture. For now I see that **Wakaama** return a `4.13 Request Entity Too Large` with `Block Size: 128` but no negotiation was done. Currently this is not clear to me what is this "Block Size" option. I was expected something like `size1` option or `block1 szx`  option :thinking:  (need to investigate)

**Comment by sbernard31** [comment]
About negotiation, I looked at californium code and it expect : block1 option with szx.
In the capture wireshark display `Block Size: 128` which is not so clear but it seems the option type is **28** which is `Size2` option.
I think it's not OK as `Size2` is for response ( see https://tools.ietf.org/html/rfc7959#section-4)
If I correctly understand in this case Wakaama want to say to Leshan : "Please use a smaller block size"
Reading the spec, I think it should send `block1 with szx1` (`size1` is for negotiation the whole payload size) : 
> Finally, a 4.13 response to a request with a Block1 Option (control
   usage, see Section 2.3) where the response carries a smaller SZX in
   its Block1 Option is a hint to try that smaller SZX.

(source : https://tools.ietf.org/html/rfc7959#section-2.9.3)
(About block1 negotiation see discussion : https://mailarchive.ietf.org/arch/msg/core/tyEr_LY_jHquT3SqAIGQbi5P7dI/)

**Comment by tuve** [comment]
I will have a look at this, the block 2 option shouldn't be set, it should be a size1 option.

**Comment by sbernard31** [comment]
@tuve, I think size1 is not OK if you want to negotiate block size. (size1 is for the whole payload size)

**Comment by tuve** [comment]
I was looking at this from rfc 7959 section 2.9.3

> also recommends that this response SHOULD include
> a Size1 Option (Section 4) to indicate the maximum size of request
> entity the server is able and willing to handle

**Comment by tuve** [comment]
Section 4 reads 

> Similarly, the Size1 Option may be used for two purposes:
> 
>    o  In a request carrying a Block1 Option, to indicate the current
>       estimate the client has of the total size of the resource
>       representation, measured in bytes ("size indication").
> 
>    o  In a 4.13 response, to indicate the maximum size that would have
>       been acceptable [RFC7252], measured in bytes.
> 

https://tools.ietf.org/html/rfc7959#section-4

**Comment by sbernard31** [comment]
The specification is maybe a bit unclear but read : https://mailarchive.ietf.org/arch/msg/core/tyEr_LY_jHquT3SqAIGQbi5P7dI/
which is clearer.
> I would expect that the server should send the desired block size if it wants the client to change from the block size it used, so 4.13 + Block1.  Size1 is not about block size, but about the size of the whole body - if the server has a limitation there, it might want to error out on a 4.13 (without a Block1 option).

**Comment by sbernard31** [comment]
So size 1 is about the whole payload (before we split it in block)
and block1+szx is to negotiation block size.

**Comment by tuve** [comment]
As I read the rfc 7959 the response should be 4.13 + size1, which in this context is "to indicate the maximum size that would have been acceptable [RFC7252], measured in bytes." as stated in section 4.

Now however the response is a size2 option which is clearly wrong.

**Comment by tuve** [comment]
block1 + szx is appropriate if thee is 2.xx response I think

**Comment by sbernard31** [comment]
@tuve should we negotiate the whole payload or the block size here ?

**Comment by tuve** [comment]
It is a actually the block size. I think the spec is a bit confusing but size1 has dual purpose. The normal way in a block transfer is to accept the block and respond with a block1 + szx to negotiate a different block size, however with a 4.13 the block is not accepted and the size 1 is the maximum size that can be handled. I guess this makes more sense (not as confusing) if there is no support for block transfer. However as I understand it we are not in a block transfer context here the client (leshan) has only sent a message without any block transfer.

Just saw you comment, that changes it bit

**Comment by tuve** [comment]
I agree if we are responding to a block1 with a 4.13 it should be block 1 + szx :) at least thats how I interpret (and I guess that was your point :D)

> Finally, a 4.13 response to a request with a Block1 Option (control
>    usage, see Section 2.3) where the response carries a smaller SZX in
>    its Block1 Option is a hint to try that smaller SZX. 

If there was not a block 1 request then it should be 4.13 + size 1, and then size 1 is a hint to retry with a block 1 where size 1 is the maximum handled (block) size. @sbernard31 do you agree with this?

**Comment by sbernard31** [comment]
> However as I understand it we are not in a block transfer context here the client (leshan) has only sent a message without any block transfer.

Nope, this is not the case. Here Leshan try to send a request (PUT) using block1 with block size 512.
We agree that we want to negotiate the block size. (this makes sense because https://github.com/eclipse/wakaama/issues/510#issuecomment-740966671 confirm that if block size 128 was used this will work, I suppose there is a limited buffer size in wakaama which is OK)

My understanding of the specification is that wakaama should answer with a 4.13 with block1+szx 128.
This is confirmed by this mail : https://mailarchive.ietf.org/arch/msg/core/tyEr_LY_jHquT3SqAIGQbi5P7dI/ from Carsten Bormann which is the author of the RFC.
> I would expect that the server should send the desired block size if it wants the client to change from the block size it used, so 4.13 + Block1. 

It also confirms that size1 is about the whole payload before we split it in block : 
>  Size1 is not about block size, but about the size of the whole body - if the server has a limitation there, it might want to error out on a 4.13 (without a Block1 option).

**Comment by sbernard31** [comment]
> I agree if we are responding to a block1 with a 4.13 it should be block 1 + szx 

We agree :relieved:  :+1: 

> If there was not a block 1 request then it should be 4.13 + size 1, and then size 1 is a hint to retry with a block 1 where size 1 is the maximum handled (block) size.

I think this is possible way and it was my understanding of the specification at the begining.
But as it sounded a bit silly to me because server should know the desired block size and so should ask for this size directly.
I finally asked [this](https://mailarchive.ietf.org/arch/msg/core/BfOlcS_BpNXbksfjzz6iGpEa_SQ/) to ietf **core** working group : 
> Now Imagine a client which sends a normal (not block1) request with 
a too large payload and server want this payload be sent by block1.
>  2) What exactly should send the server ? 4.13 or 4.13 + block1 with 
szx option ? is size1 should be used to guess the correct block size ?

And I get the [answer above](https://mailarchive.ietf.org/arch/msg/core/tyEr_LY_jHquT3SqAIGQbi5P7dI/) from Carsten which make me think that block1+szx could(should?) be used. (unless Carsten doesn't understand my question correctly :confused:

**Comment by tuve** [comment]
This was not how I read the rfc, but I must admit it makes more sense. Thanks for the clarifying links to your conversation with ieft core working group. the  I will rework this with a 4.13 + block1 + szx under normal circumstances. And 4.13 + size1, only if the response really is too big and can't be handled.

**Comment by tuve** [comment]
@sbernard31 When I am serializing a block1 option I also need to set the m bit (more) and block num. None of these makes sense if we are not in a block1 context where we reply to a block 1 request. As far as I can tell these two are only discussed in the context of a 2.xx response. I also noted this in section 2.5 of the rfc 7959. 

The error code 4.13 (Request Entity Too Large) can be returned at any
time by a server that does not currently have the resources to store
blocks for a block-wise request payload transfer that it would intend
to implement in an atomic fashion.  (Note that a 4.13 response to a
request that does not employ Block1 is a hint for the client to try
sending Block1, and a 4.13 response with a smaller SZX in its Block1
Option than requested is a hint to try a smaller SZX.)

The first part is of course not what we discuss here, but the second part, even though it does not directly contradict what Carsten answered you, it does imply that a block 1 option is not part of the response if the request wasn't a a block1 request.

Maybe the response should be

- block 1 request: 4.13 + block1 option (num = 0 (mirrored from the request but it should be 0 as this only happens for the first reques i guess), m = true, prefered szx) 
- not a block 1 request: 4.13

As I interpret 7959 this would be the correct response, but the end result could be two consecutive 4.13 to make a successful transfer, first a 4.13 to hint a retry with block transfer and then a 4.13 + block 1 option to hint retry block transfer with a different szx.

I could of corse respond  4.13 + block1 option (num = 0, m = true, szx) to a non block transfer request, but I feel this is not right, there is no block option I am responding to and no block 0. Despite the incorrectness of block num 0, I actuallty think this is a better solution but I cant escape a nagging feeling that this is a deviation from what the rfc actually says.

Should we run this through ieft a second time to make sure we get it right?

**Comment by tuve** [comment]
Now I have sleept on it :). I will go with the 4.13 + block1 option (num = 0, m = 1 and szx = max block size) regardless if it is a block transfer or not.

I am still not convinced this is correct, in fact I still think it is not what the rfc says, but I understand this is what Carsten responded and how it is implemented californium.

If this turns out to be a missinterpretation I think a block1 option should not break anything and will likely be ignored by any coap client not expecting it in the response. So I will implement with the block1 option in a response but not expect it to be present in any response given.

**Comment by sbernard31** [comment]
I understand your headache, I got pretty much the same one when I try to implement this for californium. :sweat_smile: 

> When I am serializing a block1 option I also need to set the m bit (more) and block num. None of these makes sense if we are not in a block1 context where we reply to a block 1 request.

I think you should not worry too much about this, because we have pretty much the same thing with [pre-negotiation of block2](https://tools.ietf.org/html/rfc7959#section-2.4). 
```
   To influence the block size used in a response, the requester MAY
   also use the Block2 Option on the initial request, giving the desired
   size, a block number of zero and an M bit of zero
```


> Should we run this through ieft a second time to make sure we get it right?

No problem to me if you want to get more clarification about that. But keep in mind that getting an answer is not guarantee or sometime takes a long time. So I don't advice to wait the answer to finish #461. (I mean we choose a solution and eventually we change it later, if needed)

> Now I have sleept on it :). I will go with the 4.13 + block1 option (num = 0, m = 1 and szx = max block size) regardless if it is a block transfer or not.

I think it's OK.

> If this turns out to be a missinterpretation I think a block1 option should not break anything and will likely be ignored by any coap client not expecting it in the response. So I will implement with the block1 option in a response but not expect it to be present in any response given (I will of course look for it and if present use it).

It makes sense.

> I am still not convinced this is correct, in fact I still think it is not what the rfc says, but I understand this is what Carsten responded and how it is implemented californium.

I agree it is not so clear ... 
I think you right to not accept easily to do something which seems to you "out of RFC". 
Unfortunately, in my experience, this is not the first time a RFC sounds not so clear to me. Sometime we even find thing which looks to contradiction (e.g. : https://mailarchive.ietf.org/arch/msg/tls/rQ3El3ROKTN0rpzhRpJCaKOrUyU/)
So we need to deal with this ambiguity :disappointed:

**Comment by qleisan** [comment]
Repeating the test (first post in this thread) on the latest available commit on **blocktransfer2** branch in **husqvarnagroup/wakaama** https://github.com/husqvarnagroup/wakaama/commit/adebc24eea7e2162b149989d5559bf6ddfc375eb

> While testing #461 (block transfer)

NB the PR need to be updated to the **blocktransfer2** branch
> Test with leshan server and wakaama client (leshan setting: multi-value: SENML_JSON, single-value: SENML-JSON)
> 
> "GUI write" to 31024/10/5 with ~200 bytes => "Request Entity Too Large" (leshan tries to use 512 byte block size?)
> 

This time the write is successful. Wakaama client initially responds with "Request Entity Too Large" but the data is sent again and split into two PUT requests. See wireshark log.
> "GUI read" from 31024/10/5 with ~200 bytes => OK
> 128 byte block size used. (setup with "change /31024/10/5 <200bytes>")
> 
> "GUI read" from 31024/10/5 with 1024 bytes => "Internal server error"

(this post is a work in progress - will update soon)

**Comment by tuve** [comment]
I can't replicate those errors. Tried it up with 2048 bytes. @sbernard31 and @sbertin-telular I think this is ready to merge, should fix #510

**Comment by qleisan** [comment]
@sbernard31 @tuve  "Request Entity Too Large" has been solved in #461. A new issue has been created for  "Internal server error" #526 since it was not a block transfer issue. Can I close this issue? (or are there leftovers from the conversation above)

**Comment by tuve** [comment]
I think we can close this one
